graphql_compiler/cost_estimation/cardinality_estimator.py

```diff
@@ -69,76 +69,174 @@ def _get_last_edge_direction_and_name_to_location(location):
     return edge_direction, edge_name
 
 
-def _get_parent_and_child_name_from_edge(schema_graph, child_location):
-    """Get the base classname of a location and its parent from the last edge information."""
-    edge_direction, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
+def _get_base_class_names_of_parent_and_child_from_edge(schema_graph, current_location):
+    """Return the base class names of a location and its parent from last edge information."""
+    edge_direction, edge_name = _get_last_edge_direction_and_name_to_location(current_location)
     edge_element = schema_graph.get_edge_schema_element_or_raise(edge_name)
     if edge_direction == INBOUND_EDGE_DIRECTION:
-        parent_name_from_edge = edge_element.base_out_connection
-        child_name_from_edge = edge_element.base_in_connection
+        parent_base_class_name = edge_element.base_out_connection
+        child_base_class_name = edge_element.base_in_connection
     elif edge_direction == OUTBOUND_EDGE_DIRECTION:
-        parent_name_from_edge = edge_element.base_in_connection
-        child_name_from_edge = edge_element.base_out_connection
-    return parent_name_from_edge, child_name_from_edge
+        parent_base_class_name = edge_element.base_in_connection
+        child_base_class_name = edge_element.base_out_connection
+    else:
+        raise AssertionError(u'Expected edge direction to be either inbound or outbound.'
+                             u'Found: edge {} with direction {}'.format(edge_name, edge_direction))
+    return parent_base_class_name, child_base_class_name
 
 
-def _estimate_children_per_parent(
-    schema_graph, statistics, query_metadata, parameters, child_location, parent_location
+def _query_statistics_for_vertex_edge_vertex_count(
+    statistics, query_metadata, parent_location, child_location
 ):
-    """Estimate the number of edges per parent_location that connect to child_location vertices.
+    """Query statistics for the count of edges connecting parent and child_location vertices.
 
-    Given a parent location of type A and child location of type B, assume all AB edges are
-    distributed evenly over A vertices, so the expected number of child edges per parent vertex is
-    (# of AB edges) / (# of A vertices). If A and B are subclasses of C and D respectively, we only
-    have access to CD edges, so in general, we'll use (# of CD edges) / (# of C vertices), but since
-    we're only interested in edges to Bs, we'll scale this result by the fraction of Bs over Ds.
+    Given a parent location and a child location, there are three constraints on each edge directly
+    connecting the two:
+    1. The edge class must be the same as the target location's last traversed edge.
```

Comment:
This is not as much a constraint as a specification of which edge class we will be using in this function. Also, the other two constraints are already mentioned in the args section of the docstring.

Reply:
I'm unsure how to communicate the first constraint as a specification elsewhere. Could you help out?

The args section says the following:
```
parent_location: BaseLocation, corresponding to the location the edge traversal begins from.
child_location: BaseLocation, child of parent_location corresponding to the location the
                edge traversal ends at.
```
This communicates the constraint in a very implicit fashion. Just by reading the quoted part, I wouldn't get that it's a constraint on the edge. Explicit is better than implicit.

Reply:
I'd agree with @ReaLNeroM, explicit is better than implicit and the args text is nice but insufficient. The extra few lines are worth it here.